packages/integration-tests/src/test-project-setup/data_access_from_function_project.ts

```diff
@@ -0,0 +1,195 @@
+    '../../src/test-projects/data_access_from_function';
+
+  readonly sourceProjectAmplifyDirSuffix = `${this.sourceProjectDirPath}/amplify`;
+
+  readonly sourceProjectAmplifyDirURL: URL = new URL(
+    this.sourceProjectAmplifyDirSuffix,
+    import.meta.url
+  );
+
+  /**
+   * Create a test project instance.
+   */
+  constructor(
+    name: string,
+    projectDirPath: string,
+    projectAmplifyDirPath: string,
+    cfnClient: CloudFormationClient,
+    amplifyClient: AmplifyClient,
+    private readonly lambdaClient: LambdaClient = new LambdaClient(
+      e2eToolingClientConfig
+    ),
+    private readonly cognitoIdentityProviderClient: CognitoIdentityProviderClient = new CognitoIdentityProviderClient(
+      e2eToolingClientConfig
+    ),
+    private readonly resourceFinder: DeployedResourcesFinder = new DeployedResourcesFinder()
+  ) {
+    super(
+      name,
+      projectDirPath,
+      projectAmplifyDirPath,
+      cfnClient,
+      amplifyClient
+    );
+  }
+
+  override async assertPostDeployment(
+    backendId: BackendIdentifier
+  ): Promise<void> {
+    await super.assertPostDeployment(backendId);
+
+    const clientConfig = await generateClientConfig(backendId, '1.1');
+    if (!clientConfig.data?.url) {
+      throw new Error('Data and function project must include data');
+    }
+    if (!clientConfig.data.api_key) {
+      throw new Error('Data and function project must include api_key');
+    }
+
+    // const dataUrl = clientConfig.data?.url;
+
+    const httpLink = new HttpLink({ uri: clientConfig.data.url });
+    const link = ApolloLink.from([
+      createAuthLink({
+        url: clientConfig.data.url,
+        region: clientConfig.data.aws_region,
+        auth: {
+          type: AUTH_TYPE.API_KEY,
+          apiKey: clientConfig.data.api_key,
+        },
+      }),
+      // see https://github.com/awslabs/aws-mobile-appsync-sdk-js/issues/473#issuecomment-543029072
+      httpLink,
+    ]);
+    const apolloClient = new ApolloClient({
+      link,
+      cache: new InMemoryCache(),
+    });
+
+    await this.assertDataFunctionCallSucceeds(apolloClient);
+    await this.assertNoopWithImportCallSucceeds(apolloClient);
+  }
+
+  private assertDataFunctionCallSucceeds = async (
+    apolloClient: ApolloClient<NormalizedCacheObject>
+  ): Promise<void> => {
+    const response = await apolloClient.query<number>({
+      query: gql`
+        query todoCount {
+          todoCount
+        }
+      `,
+      variables: {},
+    });
+
+    assert.deepEqual(response.data, { todoCount: 0 });
+  };
+
+  private assertNoopWithImportCallSucceeds = async (
+    apolloClient: ApolloClient<NormalizedCacheObject>
+  ): Promise<void> => {
+    const response = await apolloClient.query<number>({
+      query: gql`
+        query noopImport {
+          noopImport
+        }
+      `,
+      variables: {},
+    });
+
+    assert.deepEqual(response.data, { noopImport: 'STATIC TEST RESPONSE' });
+  };
+}
```

Comment:
I don't seem to see any code that would call the lambda.
If the feature is about lambda accessing appsync, we should be invoking lambda and asseserting that lambda intracted with appsync.

The test as is doesn't prove that lambda is working.

Some ideas how to test lambda can be found [here](https://github.com/aws-amplify/amplify-backend/blob/d227f96793fd6a3e181af96bbb1ef2f0d3ab7ae4/packages/integration-tests/src/test-project-setup/data_storage_auth_with_triggers.ts#L217-L241) - we assert that lambda interacts with s3 bucket there, similar strategy can be applied here for appsync.

Reply:
Both of these tests call a query, which each [call a lambda](https://github.com/aws-amplify/amplify-backend/pull/2224/files#diff-6ab6d1ba968c8f4bd0c0f18c3c6ab9f5c80cc036055ed09b00bd3a1a4bb21f32R13-R25) and the response assertions are specific to the lambda being called ([todoCount](https://github.com/aws-amplify/amplify-backend/pull/2224/files#diff-2807ad44b84ffd9d18ec3423889231abb0a54bf424b0be74c3295685625c94d3R19) and [noopImport](https://github.com/aws-amplify/amplify-backend/pull/2224/files#diff-623f11d82c5ce3d595e11768cfbcaf371e3405fb30668a217bd76eea61ade840R21)).

I considered testing the lambda's in isolation (without wiring these calls through AppSync query calls), but that use-case is a simpler subset of the use-cases covered by these tests. 

Can you help me understand the coverage gap that needs to be filled?

Reply:
I see. There's no coverage gap then.

Can you add JSDoc comment to both `private assert...` methods explaining that they indeed call the lambda and how ? For the reader it's not obvious that GQL calls translate to lambda invocations here.

Reply:
Understood. JSDoc added